### PR TITLE
[IMP] core: add session deletion mechanism

### PR DESCRIPTION
**Pull request by thle-odoo**
Objective:
----------
A user must be able to see which of his sessions/devices are active.
Make it easy to block devices and analyse current sessions.
If a user notices unusual operations concerning him on another device,
he must be able to stop these operations by blocking the session used by this device.

General:
--------
A device is uniquely identified by its name, IP address and the session to which it is linked.
Consequently, a device is always linked to a session, and a session is always linked to at least one device.

A device is tracked with its last activity.
The same device is updated every X seconds in order to track its use in terms of duration of activity.

Tracking data is collected within the session itself.
This data is updated in the database when necessary (according to elapsed time).

Within the database, data are inserted as logs (model `res.device.log`).
The `res.device` model is responsible for filtering the logs so that the user can obtain usable data.
Logs are cleaned during garbage collection.

Each user sees data from their own devices.
Administrators can view all devices.

Blocking sessions:
------------------
Blocking a session by selecting a device must block the session directly.
From an administrator's point of view, if we want to block a session,
we expect the session file to be deleted directly on the filesystem.

If the session file is no longer present on the filesystem, we can be sure
that there is no longer any risk of session usurpation.

Note:
There are several ways of blocking the session, but this is the safest.
Blocking a session based on DB values does not cover scenarios in which we perform a backup, for example.

Find the session file:
----------------------
To find a session file on the filesystem, we need to know the sid of the session (as this is its filename).
We don't want to store the sid in the database.
However, we can store a part of it with:
- a large enough part to be certain of the uniqueness of the session;
- a small enough part that we cannot brute force the end of the sid.

Browsing the filesystem has a certain performance cost (and can therefore have defects if abused).

The proposed solution is to change the granularity of the way we store the session on the filesystem.
This means finding a compromise between sub-folders and files per sub-folder to browse for a session file.

The sid will be Base64 encoded in order to increase the number of sub-folders (64^2 instead of 16^2).

Task:3627898

### CODE REVIEW

File: odoo/addons/base/views/res_users_views.xml
Comment (beledouxdenis) [inline odoo/addons/base/views/res_users_views.xml:615]:
```suggestion
                                        <t t-out="luxon.DateTime.fromISO(record.last_activity.raw_value).toRelative()"/>
```
You might need to ask a JS framework expert if this is allowed, because I did not see a single occurence in the code to this one exactly, but that works and it avoids the compute field `last_activity_display`

File: odoo/addons/base/models/res_device.py
Comment (beledouxdenis) [inline odoo/addons/base/models/res_device.py:128]: Nice !

**Review by tbe-odoo — CHANGES_REQUESTED** [review]
Hello, thanks for the work done on your PR.

I have a few suggestions we already talked about but haven't yet submitted my review.
No big issues in your PR, we just need to improve a little bit the code readability. 

Don't worry or rush out, the database scheme should be kept this way, so for the freeze it won't be blocking if you apply those changes after :) 

Good work 

Edit: By the way, you should take a look at the `CI/style` errors

File: odoo/addons/base/models/res_device.py
Comment (tbe-odoo) [inline odoo/addons/base/models/res_device.py:42]:
```suggestion
            device.display_name = f"{platform.capitalize()} {browser.capitalize()}"
```

File: odoo/addons/base/models/res_device.py
Comment (tbe-odoo) [inline odoo/addons/base/models/res_device.py:66]:
```suggestion
        return platform.lower() in mobile_platform
```

File: odoo/addons/base/models/res_device.py
Comment (tbe-odoo) [inline odoo/addons/base/models/res_device.py:122]:
```suggestion
                    AND log1.platform = log2.platform
                     AND log1.browser = log2.browser
```

File: odoo/service/security.py
Comment (seb-odoo) [inline odoo/service/security.py:16]: Hello, couldn't this method be made to only receive `request` as other params can be read directly from it?
It seems to never be called without a `request`.

Comment (beledouxdenis) [reply odoo/service/security.py:16]: It is called without it, in one place, in the bus or something if I remember correctly. I had the same comment and then saw that extreme case in one place where it was called without it and without possibility to call it with it.

Comment (seb-odoo) [reply odoo/service/security.py:16]: In `_dispatch_bus_notifications`, called in `get_messages`, called in `_serve_forever` where a `WebsocketRequest` is created on the next line.
I wonder if we could move the creation of the request one line up to avoid this limitation.
cc @tsm-odoo

Comment (beledouxdenis) [reply odoo/service/security.py:16]: I didn't want to take the risk now for this task, as it was a bit out of scope of this, but do it in a next iteration.

File: odoo/addons/base/models/res_device.py
Comment (ryv-odoo) [inline odoo/addons/base/models/res_device.py:49]: Missing domain, it won't scale. `[('session_identifier', 'in', self.mapped('session_identifier')]` maybe ?

File: odoo/addons/base/models/res_device.py
Comment (ryv-odoo) [inline odoo/addons/base/models/res_device.py:147]: Nitpicking: Not really important but it is not deterministic to use a set because of the PYTHONHASHSEED. => `list(unique(device.session_identifier for device in self))`

**Review by ryv-odoo — COMMENTED** [review]
Hi @thle-odoo ,

Nice job and nice feature. I added few suggestions, don't hesitate if you have question.

File: odoo/addons/base/models/res_device.py
Comment (thle-odoo) [reply odoo/addons/base/models/res_device.py:31]: In fact, the index is useless in this case because the number of `False` values is much greater than the number of `True` values and this field is only used in a `where` condition to filter with `False` values, postgresql will not use the index. Thanks for this comment !

File: odoo/addons/base/models/res_device.py
Comment (thle-odoo) [reply odoo/addons/base/models/res_device.py]: Yes, we only need records that respect this domain.

File: odoo/addons/base/models/res_device.py
Comment (beledouxdenis) [reply odoo/addons/base/models/res_device.py]: Well it is debatable, and we got the debate in the security team.

One of the goal is to detect when you get your session stolen by an attacker, and if:
- The attacker only uses readonly routes,
- You no longer yourself use the session that got stolen
The information will never be sent to the database, and you will never see that your session got stolen.

So we need to choose between performance and security in this case. For the moment we chose security, but we can consider prioritizing performance.

File: odoo/service/security.py
Comment (beledouxdenis) [reply odoo/service/security.py:16]: Then you do not do that `_update_device` for websocket calls.
We analyzed the stack of all possible ways where a session can be used, and this `check_session` is the common place for all ways: HTTP calls, websocket calls, XMLRPC calls.

**Comment by tbe-odoo** [comment]
@beledouxdenis @thle-odoo I was asked if we want to merge this PR for 17.4, if so, we should do it asap since our prod will be migrated directly to this version and it should be tested in our prod env.

I see a few changes suggested by JUC, I feel like we're not that far from merging this PR. What do you think?
Is it possible to do the changes for this week?

File: odoo/http.py
Comment (thle-odoo) [reply odoo/http.py:1117]: We can assume that these keys will always be present in the `trace` and therefore use `[]` (I will change it).

File: odoo/addons/base/models/res_device.py
Comment (thle-odoo) [reply odoo/addons/base/models/res_device.py]: My opinion is to prioritize security.
If we notice that this has an impact on performance, we can improve it in the future (by adding a flag to the trace which determines whether it is synchronized and insert all non synchronized trace in a session when the cursor is not read only) at the expense of security.

File: odoo/http.py
Comment (mart-e) [inline odoo/http.py:985]: should we check that we have a properly formed identifier? e.g. could I bruteforce all 2 letters identifiers to delete every sessions

Comment (tbe-odoo) [reply odoo/http.py:985]: A check for the identifier length should be added.

Comment (thle-odoo) [reply odoo/http.py:987]: Thanks for this remark !

File: odoo/http.py
Comment (tbe-odoo) [inline odoo/http.py:983]:
```suggestion
            #  avoid to remove a session if less than 42 chars
            #  This prevent malicious users to delete sessions from a different database
            # by specifying a `res.device.log` with only 2 characters
            if len(identifier) < 42:
```

**Comment by tbe-odoo** [comment]
robodoo override=ci/security

File: odoo/service/security.py
Comment (Julien00859) [reply odoo/service/security.py:16]: ah yes, websocket, indeed

**Pull request closed without merging**
